### PR TITLE
Rewrite poutine.do to represent single-world counterfactuals

### PR DESCRIPTION
**Pull request by eb8680**
Addresses #2153.

Before this PR, `poutine.do` and `poutine.condition` did not compose in an intuitive way to produce counterfactual queries.  After the changes in this PR, `poutine.do` modifies models to represent the joint distribution over a variable to be intervened on and a potential outcome after a particular intervention on that variable, as opposed to simply mutilating the model.

cc @robertness - what do you think of this change (proposed in #2153)?  Is it sensible?  Would it break a bunch of existing code?  I'm far from being an expert in causal inference but I'm hoping to find ways we can automate away more do-calculus drudgery in Pyro.

### CODE REVIEW

File: pyro/poutine/do_messenger.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    :returns: stochastic function decorated with a :class:`~pyro.poutine.do_messenger.DoMessenger`
+    """
+    def __init__(self, data):
+        super(DoMessenger, self).__init__()
+        self.data = data
+        self._intervener_id = str(id(self))
+
+    def _pyro_sample(self, msg):
+        if msg.get('_intervener_id', None) != self._intervener_id and \
+                msg['name'] in self.data and \
```
Comment (fritzo) [inline pyro/poutine/do_messenger.py]: nit: simplify to `self.data.get(msg['name']) is not None`

File: tests/poutine/test_counterfactual.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
+import logging
+
+import pytest
+
+import pyro
+import pyro.distributions as dist
+import pyro.poutine as poutine
+from tests.common import assert_equal, assert_not_equal
+
+logger = logging.getLogger(__name__)
```
Comment (fritzo) [inline tests/poutine/test_counterfactual.py]: logger is unused, causing flake8 failure

**Comment by eb8680** [comment]
> Do I understand correctly

Yes, I believe so, except that

> Before this PR poutine.do() would block the guide from conditioning the model, and the model would erroneously draw x from the prior

is not quite right: under the current behavior the sample site `"x"` would return `intervention` in both model and guide but would not appear in either trace.

File: pyro/poutine/do_messenger.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+            intervention = self.data[msg['name']]
+            msg['name'] = msg['name'] + "__CF"  # mangle old name
+
+            # if intervention is None, use same variable with different randomness
+            # this is equivalent to the posterior predictive distribution
+            if intervention is None:
+                intervention = msg['fn']
+
+            if isinstance(intervention, Distribution):
+                # stochastically intervened site is no longer observed
```
Comment (fritzo) [inline pyro/poutine/do_messenger.py]: Maybe we should leave the `Distribution` case `NotImplemented` for the release?

It seems more subtle and is untested. For example in the variational intervention case
```py
with poutine.do(...):
    with poutine.trace() as tr:
        guide()
    with poutine.replay(trace=tr.trace):
        model()
```
it seems that the samples drawn from the distributional intervention will not agree between the guide and the model.

**Pull request merged**